### PR TITLE
2.7: New password-based authenticated encryption sample program pbcrypt

### LINKED ISSUE #1906: aescrypt2 shows bad practice
The very first sample program in alphabetical order, `programs/aes/aescrypt2.c`, exhibits several bad practices. It rolls its own CBC mode on top of ECB. It does AEAD by combining AES-CBC with HMAC (at least it does encrypt-then-mac). It rolls its own key stretching by iterating SHA-256.

We should replace this by a program that illustrates best practices, and possibly by one or more other programs that illustrate how to use the low-level functions but not to implement dodgy schemes.

This is a bug, even though the program works, because people who don't know better may copy the code, and it gives a bad impression of the library to people who do know better.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
This PR introduces `programs/cipher/pbcrypt`, a new sample program to demonstrate password-based authenticated encryption.

This is meant to illustrate good practice, so please review accordingly. A few caveats:

* The program works in a specific way: streaming AEAD. Since the library doesn't support streaming CCM, the program doesn't support it.
* The use of PKCS#12 key derivation is hard-coded because the library doesn't have a generic interface to password-based key derivation.
* The program should build in all configurations for which the library builds, but it may not run usefully in some “exotic” configurations (e.g. GCM enabled but no block cipher that supports GCM).
* The program relies on argv and stdio. It does password-based key derivation so it isn't meant for microcontrollers anyway.

Provide a demo usage script and run it from `all.sh`.

Remove the badly designed `aescrypt2`. Fix #1906.

This PR is for 2.7. I'll forward-port it to 2.16+ once it's approved. I think this will mean the following changes;
* Add ChachaPoly to the `#ifdef` about `main` and to `cipher_is_aead`.
* Replace `mbedtls_zeroize` by `mbedtls_platform_zeroize`.
* Even in development, this is a sample program for the mbedtls API. In the PSA API, there's no password-based key derivation yet and multi-part AEAD is not implemented yet, but `key_ladder_demo` demonstrates symmetric key derivation and single-part AEAD.

### CODE REVIEW

**Comment by mpg** [comment]
@gilles-peskine-arm  There appears to be a significant intersection with #2698 - adding sample usage scripts for programs and testing them in the CI. Does that mean the two PRs conflicts with each other? If that's the case, I'd suggest giving priority to #2698 and labeling this one as "needs: preceding PR".

**Comment by gilles-peskine-arm** [comment]
The intersection was only adding something to `all.sh`. I've removed that part: #2698 is a better place for it since that's its main job. Now the PRs are fully independent, except insofar as the format of `pbcrypt_demo.sh` only makes full sense in combination with #2698.

**Comment by mpg** [comment]
Ok, thanks!

File: programs/cipher/pbcrypt_demo.sh [hunk 1]
```diff
@@ -0,0 +1,87 @@
+#!/bin/sh
+set -e -u
```
Comment (gilles-peskine-arm) [inline programs/cipher/pbcrypt_demo.sh:2]: When #2698 is ready and backported, include `demo_common.sh` and adapt the code here accordingly.

File: programs/cipher/pbcrypt.c [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+                                 "fread: short read at line %d\n",      \
+                                 __LINE__ );                            \
+                ret = PBCRYPT_ERR;                                      \
+                goto exit;                                              \
+            }                                                           \
+        }                                                               \
+        while ( 0 )
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
```
Comment (mpg) [inline programs/cipher/pbcrypt.c:153]: Note: in the meantime we introduced `mbedtls_platform_zeroize()`.

Comment (gilles-peskine-arm) [reply programs/cipher/pbcrypt.c:153]: Not in 2.7, and already noted for the other branches (see PR description).

Comment (mpg) [reply programs/cipher/pbcrypt.c:153]: Oh, right. Sorry for the noise then.

File: programs/cipher/pbcrypt.c [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+                                        (unsigned long long) header[1] << 48 |
+                                        (unsigned long long) header[2] << 40 |
+                                        (unsigned long long) header[3] << 32 |
+                                        (unsigned long long) header[4] << 24 |
+                                        (unsigned long long) header[5] << 16 |
+                                        (unsigned long long) header[6] << 8 |
+                                        (unsigned long long) header[7] ),
+                       payload_size <= SIZE_MAX,
+                       "Payload too large" );
+    metadata->payload_size = payload_size;
```
Comment (dgreen-arm) [inline programs/cipher/pbcrypt.c:313]: This line is causing failures on Windows due to the conversion from `unsigned long long` to `size_t`. Given the check above it should be fine, but the compiler is complaining about it.

**Comment by gilles-peskine-arm** [comment]
Due to the imminent retirement of 2.7, I am closing this pull request. I have ported the program to 2.16 in https://github.com/ARMmbed/mbedtls/pull/4216. Compared to the 2.7 version, I added ChaChaPoly support and fixed a couple of minor bugs, all in new commits.

**Pull request closed without merging**